test/common/tmpdir.js

```diff
@@ -31,28 +31,56 @@ function refresh() {
   }
 }
 
+function logErrorInfo(err) {
+  console.error('Can\'t clean tmpdir:', tmpPath);
+
+  const files = fs.readdirSync(tmpPath);
+  console.error('Files blocking:', files);
+
+  if (files.some((f) => f.startsWith('.nfs'))) {
+    // Warn about NFS "silly rename"
+    console.error('Note: ".nfs*" might be files that were open and ' +
+                  'unlinked but not closed.');
+    console.error('See http://nfs.sourceforge.net/#faq_d2 for details.');
+  }
+
+  console.error();
+
+  // Additionally logging err, just in case throwing err gets overshadowed by a
+  // test failure error.
+  console.error(err);
+
+  throw err;
+}
+
 function onexit() {
   // Change directory to avoid possible EBUSY
   if (isMainThread)
     process.chdir(testRoot);
 
-  try {
-    rmSync(tmpPath);
-  } catch (e) {
-    console.error('Can\'t clean tmpdir:', tmpPath);
+  const maxRetries = 10;
+  for (let retryNumber = 0; retryNumber < maxRetries; ++retryNumber) {
+    let err = null;
 
-    const files = fs.readdirSync(tmpPath);
-    console.error('Files blocking:', files);
+    try {
+      rmSync(tmpPath);
+    } catch (_err) {
+      err = _err;
+    }
 
-    if (files.some((f) => f.startsWith('.nfs'))) {
-      // Warn about NFS "silly rename"
-      console.error('Note: ".nfs*" might be files that were open and ' +
-                    'unlinked but not closed.');
-      console.error('See http://nfs.sourceforge.net/#faq_d2 for details.');
+    if (err === null) {
+      break;
     }
 
-    console.error();
-    throw e;
+    logErrorInfo(err);
+
+    const errPath = err.path;
+    const errCode = err.code;
+
+    if (errCode === 'EACCES' || errCode === 'EPERM') {
+      const surroundingDir = path.join(errPath, '..');
+      fs.chmodSync(surroundingDir, 0o777);
```

Comment:
Do we need extras checks to make sure we're not changing access to a directory outside (e.g. the parent) of the tmpdir?

Reply:
Since `err` comes from `rmSync(tmpPath);`, the `path` in `err` will always be something inside `tmpPath`, so I don't think we need more checks.